### PR TITLE
add method to user-backend to check if allowed to update email address

### LINKED ISSUE #14127: Admins can't edit mails of users if `allow_user_to_change_display_name` is set to false
Had a little chat with @enoch85 and he couldn't change the mail of users even as admin. Turned out that he had `allow_user_to_change_display_name` set to false and from the code logic admins can't change the users email then.

We either need to allow it for admins or hide those settings in the UI.

### PR DESCRIPTION
**Pull request by georgehrke**
fixes https://github.com/owncloud/core/issues/14127

please review @PVince81 @MorrisJobke

### CODE REVIEW

**Comment by karlitschek** [comment]
makes sense :-1:

File: lib/public/iuser.php [hunk 1]
```diff
@@ -130,6 +130,14 @@ public function canChangePassword();
 	public function canChangeDisplayName();
 
 	/**
+	 * check if the backend supports changing e-mail addresses
+	 *
+	 * @return bool
+	 * @since 8.2.0
+	 */
+	public function canChangeEMailAddress();
```
Comment (PVince81) [inline lib/public/iuser.php]: Will this break existing backends that might not implement it ? (ex: in third party apps that might implement this interface)

Comment (georgehrke) [inline lib/public/iuser.php]: @blizzz Was this the case when you added `public function canChangeAvatar();`?

Comment (blizzz) [inline lib/public/iuser.php]: @georgehrke I did not add it into an interface. This was added by @icewind1991. Anyway, every backend will break, which implements IUser and does not adopt in time to this change.

Comment (georgehrke) [inline lib/public/iuser.php]: So, yes, it'll break user backends, which for whatever reason implement iUser, but it's just a matter of implementing a single method that returns a boolean, so I don't think it's a big deal, as long as we put a big note into the changelog.

Comment (blizzz) [inline lib/public/iuser.php]: See all the broken update complaints? 

What about introducing and checken for a new Interface? I think we need to  switch thereto anyway in Users to be able to further extend it with optional capabilities.

Comment (DeepDiver1975) [inline lib/public/iuser.php]: bäääh .. this is all soooooooo crappy

Can we postpone this to 9.0 please?

I'd really like to introduce an accounts table which holds all user information (id, displayname, email, quota, home ...).
Backends are used for authentication and can provide additional information (groups, email, display etc) which are to be stored in the accounts table.

Benefit:
- We know all users any point in time (even e.g. backend systems are offline)
- We have one data source for all user information
- We can maintain user information in on place (an no longer need these canChange* methods)
- We can finally cleanup the lower case username query issue #19217

Sideeffect:
- Kill all the interfaces :speak_no_evil:

Comment (rullzer) [inline lib/public/iuser.php]: > Backends are used for authentication and can provide additional information (groups, email, display etc) which are to be stored in the accounts table.

Keeping that in sync will be a pain...

Comment (DeepDiver1975) [inline lib/public/iuser.php]: > Keeping that in sync will be a pain...

Less then the pain we feel today ... note that LDAP is already syncing this information.
In addition this info can be synced on login (or similar events within sso deployments)

**Comment by MorrisJobke** [comment]
Moved to 9.0 as per https://github.com/owncloud/core/issues/14127#issuecomment-144801223

File: lib/public/iuser.php (hunk 1 above)
Comment (PVince81) [inline lib/public/iuser.php]: @DeepDiver1975 is there a ticket for the account table enhancement ?

Comment (PVince81) [inline lib/public/iuser.php]: One approach would be to create a new interface (V2) with the new method, and make that interface extend this one. So "user backend v2" would implement the new interface.

Another possible alternative but possibly less popular is to use traits ? @icewind1991

Comment (PVince81) [inline lib/public/iuser.php]: Accounts table ticket is here: https://github.com/owncloud/core/issues/21282

Comment (LukasReschke) [inline lib/public/iuser.php]: Since that is 9.1 let's move that as well here?

Comment (LukasReschke) [inline lib/public/iuser.php]: … or we we want to close this one completely?

Comment (PVince81) [inline lib/public/iuser.php]: Seems there won't be an account table for 9.1 after all.

We should move forward with this or close it. No point in leaving a PR open for that long.

**Pull request closed without merging**